### DISCUSSION TITLE
support for "virtual" file metadata changes

### DISCUSSION THREAD

**Discussion by ak2k** [fsspec/filesystem_spec#1031] [Q&A] [answered]
Admittedly, this might come across as a strange goal. I am accessing a ZIP object (over https) via an fsspec filesystem that contains a number of CSV files. I would ideally like to rename/mv/move these contained CSVs within the scope of the session, but then have no attempt made to persist those changes beyond the fsspec session. In essence, I'm hoping to create a view with different file metadata on top of an existing ZIP file, for the open session only and without change to the original.

Is anything like this contemplated/possible in fsspec?

Thank you!

**Comment by martindurant** [comment, marked as answer]
The ZIP filesystem doesn't allow for this kind of thing, but maybe you could look into ReferenceFileSystem, which maps files on any filesystem backend (including ZIP) to arbitrarily map filesystem names to targets.

**Comment by ak2k** [comment]
Thank you!

Separately, I wondered whether starting a transaction that never completes
might attempt to defer the changes. Would that be the best way to try to
use ReferenceFileSystem in this way?

**Reply by martindurant** [reply]
If you are using ReferenceFS, you do not need a transaction, it never writes to the target.

**Reply by martindurant** [reply]
But I thought you wanted to map the paths for the whole session, so it's enough to set up the right mapping, you shouldn't need any write operations at all.

**Reply by ak2k** [reply]
Yes, that's perfect. I'll get going in that direction.

(My question regarding transaction was more whether that could be a means for indefinitely deferring the writes if that would make things work with another FS, but it sounds like ReferenceFS is the better approach).

**Reply by martindurant** [reply]
> a means for indefinitely deferring the writes if that would make things work with another FS

It will depend on how temporaries work on that filesystem. On most backends, it amounts to making a temporary file, and moving it to the final location when done. I'm not sure what ZIP doe offhand, but it likely needs some kind of write even while the transaction is live.

**Reply by ak2k** [reply]
Makes sense, thank you.

**Comment by martindurant** [comment]
The specification for the references is here: https://fsspec.github.io/kerchunk/spec.html
For your case, you make a JSON file (or dict) with an inner references dict of `{path: [target]}` where `path` is how the file is known within the referenceFS, and `target` is the URL you want to map to. The references are stored in the `references` attribute of the filesystem instance, so you can even edit it directly in-place, if you wish.
